### PR TITLE
Simplify ObjectAdapter

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR is a clean-up / light refactoring of ObjectAdapter.

The main change is the removal of CreateDirectProxy / CreateIndirectProxy and the addition of public get-only properties for AdapterId, ReplicaGroupId and PublishedEndpoints. If you want a direct or indirect proxy, you can simply create a proxy and then call Clone and use one of these properties.

### CODE REVIEW

File: csharp/src/Ice/ObjectAdapter.cs [hunk 1]
```diff
@@ -18,6 +18,10 @@ namespace ZeroC.Ice
     /// servants, identities, and proxies.</summary>
     public sealed class ObjectAdapter : IDisposable, IAsyncDisposable
     {
+        /// <summary>Returns the adapter ID of this object adapter, or the empty string if this object adapter does not
+        /// have an adapter ID.</summary>
+        public string AdapterId { get; }
```
Comment (externl) [inline csharp/src/Ice/ObjectAdapter.cs:23]: Wouldn't `null` make more sense for not set?

File: csharp/src/Ice/ObjectAdapter.cs [hunk 2]
```diff
@@ -40,6 +44,23 @@ public sealed class ObjectAdapter : IDisposable, IAsyncDisposable
...
+                lock (_mutex)
+                {
+                    return _publishedEndpoints;
+                }
+            }
+        }
+
+        /// <summary>Returns the replica group ID of this object adapter, or the empty string if this object adapter
+        /// does not belong to a replica group.</summary>
+        public string ReplicaGroupId { get; }
```
Comment (externl) [inline csharp/src/Ice/ObjectAdapter.cs:62]: Ditto about `null`

**Review by externl — APPROVED** [review]

**Review by pepone — APPROVED** [review]
Looks good, but I think is a bit odd that we have a getter that is a property and a setter that is a method for PublishedEndpoints, it works but seems an uncommon pattern

**Comment by bernardnormier** [comment]
> Looks good, but I think is a bit odd that we have a getter that is a property and a setter that is a method for PublishedEndpoints, it works but seems an uncommon pattern

The main setter is `SetPublishedEndpointsAsync` and `SetPublishedEndpoint` is just a helper and I don't think it should be the "set" of the PublishedEndpoints property.

And I find a get-only PublishedEndpoints property more logical than a `GetPublishedEndpoints()` method.

File: csharp/src/Ice/ObjectAdapter.cs (hunk 1 above)
Comment (bernardnormier) [reply csharp/src/Ice/ObjectAdapter.cs:23]: I didn't change this logic, just the API. 

We could switch to nullable strings for adapter IDs and replica group IDs however this affects much more code and we'd have to decide how to handle empty adapter IDs / replica group IDs - either like null or as illegal values.

File: csharp/src/Ice/ObjectAdapter.cs [hunk 3]
```diff
@@ -40,6 +44,23 @@ public sealed class ObjectAdapter : IDisposable, IAsyncDisposable
         /// determines this protocol.</summary>
         public Protocol Protocol { get; }
 
+        /// <summary>Returns the endpoints listed in a direct proxy created by this object adapter.</summary>
```
Comment (pepone) [inline csharp/src/Ice/ObjectAdapter.cs:47]: Maybe add `<see cref="SetPublishedEndpoints"/> and <see cref="SetPublishedEndpointsAsync"/>` in the summary here

Comment (bernardnormier) [reply csharp/src/Ice/ObjectAdapter.cs:47]: Done.

**Pull request merged**